packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,182 @@
+  | {
+      status: BackfillSyncStatus.Syncing;
+      target: ChainTarget;
+    }
+  | {status: BackfillSyncStatus.Idle};
+
+export type RangeSyncModules = {
+  chain: IBeaconChain;
+  db: IBeaconDb;
+  network: INetwork;
+  config: IBeaconConfig;
+  logger: ILogger;
+};
+
+export type BackfillSyncOpts = {
+  batchSize: number;
+};
+
+export class BackfillSync {
+  private readonly chain: IBeaconChain;
+  private readonly network: INetwork;
+  private readonly db: IBeaconDb;
+  private readonly config: IBeaconConfig;
+  private readonly logger: ILogger;
+  private opts: BackfillSyncOpts;
+  //last fetched slot
+  private lastFetchedSlot: Slot | null = null;
+  //last trusted block
+  private anchorBlock: SignedBeaconBlock | null = null;
+  private processor: ItTrigger = new ItTrigger();
+  private peers: PeerMap<phase0.Status> = new PeerMap();
+
+  constructor(modules: RangeSyncModules, opts?: BackfillSyncOpts) {
+    this.chain = modules.chain;
+    this.network = modules.network;
+    this.db = modules.db;
+    this.config = modules.config;
+    this.logger = modules.logger;
+    this.opts = opts ?? {batchSize: 32};
+    this.network.events.addListener(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.addListener(NetworkEvent.peerDisconnected, this.removePeer);
+    void this.init();
+  }
+
+  /** Throw / return all AsyncGenerators */
+  close(): void {
+    this.network.events.removeListener(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.removeListener(NetworkEvent.peerDisconnected, this.removePeer);
+    this.processor.end(new ErrorAborted("BackfillSync"));
+  }
+
+  async init(): Promise<void> {
+    const anchorState = this.chain.getHeadState();
+    const parentBlockRoot: Root = anchorState.latestBlockHeader.parentRoot;
+    this.anchorBlock = await this.db.blockArchive.getByRoot(parentBlockRoot);
+    void this.sync();
+  }
+
+  async sync(): Promise<void> {
+    try {
+      for await (const _ of this.processor) {
+        if (this.peers.size === 0) continue;
+
+        const oldestRequiredEpoch = Math.max(
+          GENESIS_EPOCH,
+          this.chain.clock.currentEpoch - getMinEpochForBlockRequests(this.config)
+        );
+        const oldestSlotRequired = computeStartSlotAtEpoch(this.config, oldestRequiredEpoch);
+
+        if (this.anchorBlock) {
+          try {
+            const toSlot = this.lastFetchedSlot ?? this.anchorBlock.message.slot;
+            const fromSlot = Math.max(toSlot - this.opts.batchSize, oldestSlotRequired);
+            if (fromSlot >= toSlot) {
+              this.logger.info("Backfill sync completed");
+              break;
+            }
+            await this.syncRange(fromSlot, toSlot, this.anchorBlock.message.parentRoot);
+          } catch (e) {
+            this.logger.debug("Error while backfiling by range", e);
+          }
+        } else {
+          try {
+            await this.syncBlock(this.chain.getHeadState().latestBlockHeader.parentRoot);
```

Comment:
We must backfill from the weak subjectivity point which becomes our finalized checkpoint. Shouldn't we get the finalized checkpoint instead of the head here?

Reply:
Yeah, I think that makes more sense.